openfoodfoundation/openfoodnetwork
# Reduce restrictions for adding a variant to a subscription

#### What? Why?

Closes #2243 

From @sstead:

> If I try to edit or create a susbcription when there are no future OCs setup, I can't really add any products to a subscripton and I get error messages.
>
> The logic of this control was initially designed in teh case that customers were creating their subs. It would prevent them from adding rare items to their subs, and maximise the changes that items in their sub would be available.

With this PR, adding the following variants to a subscription are now allowed:

* Variants for which the shop is the supplier
* Variants of suppliers that have granted the shop permission to "add to order cycle"
* Variants which are in an outgoing exchange to the shop, regardless of the timing of the OC

The user is now just warned if there are no open or upcoming OCs for the variant in the subscription, in the subscription form:

![20180917060705-subscription-add](https://user-images.githubusercontent.com/2243/45601591-063bf280-ba42-11e8-93da-ab8334ed7744.png)

An when reviewing subscription details:

![20180917060816-subscription-review](https://user-images.githubusercontent.com/2243/45601596-181d9580-ba42-11e8-9dea-66f3841ae936.png)

Because there are fewer restrictions now, there has been opportunity to have the `spree/admin/variants#search` (for requests from subscription form) and subscription validation share the same logic, now in `OpenFoodNetwork::SubscriptionService`.

#### What should we test?

Testing requirements:

* Subscriptions are enabled for the shop (Shop Preferences > Subscriptions)

Set up the following:

1. Product for which the shop is the supplier
2. Product with permitted supplier - Have the supplier grant "add to order cycle" permission to the shop.
3. Product in an outgoing exchange for a completed OC of the shop
4. Product in an outgoing exchange for a currently open OC of the shop
5. Product in an outgoing exchange for a future OC of the shop
6. Unrelated product

Then, to test:

1. Create a subscription for the customer.
2. Without saving the subscription, attempt to add Products 1 to 6 above. It should not be possible to add Product 6.
3. For Products 1 to 3, there should be a warning: "There are no open or upcoming order cycles for this product."
4. Save the subscription. This should be successful.
5. Edit the subscription. In the review page, Products 1 to 3 should have warning: "Unavailable"
6. Click to edit products in the subscription. There should still be warnings for Products 1 to 3.
7. Change the start date of the subscription. Saving should be successful.

#### Release notes

- Reduce restrictions for adding a variant to a subscription. The following are now allowed:
  * Variants for which the shop is the supplier
  * Variants of suppliers that have granted the shop permission to "add to order cycle"
  * Variants which are in an outgoing exchange to the shop, regardless of the timing of the OC

Changelog Category: Changed
